url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/core/RAMJobStore.java

```diff
@@ -0,0 +1,844 @@
+        orig.jobDetail = jw.jobDetail; // already cloned
+      }
+    }
+  }
+
+  /**
+   * <p>
+   * Remove (delete) the <code>{@link org.quartz.jobs.Job}</code> with the given name, and any <code>{@link org.quartz.triggers.Trigger}</code> s that
+   * reference it.
+   * </p>
+   *
+   * @return <code>true</code> if a <code>Job</code> with the given name & group was found and removed from the store.
+   */
+  @Override
+  public boolean removeJob(String jobKey) {
+
+    boolean found = false;
+
+    synchronized (lock) {
+      List<Trigger> triggers = getTriggersForJob(jobKey);
+      for (Trigger trig : triggers) {
+        this.removeTrigger(trig.getName());
+        found = true;
+      }
+
+      found = (jobsByKey.remove(jobKey) != null) | found;
+      if (found) {
+
+      }
+    }
+
+    return found;
+  }
+
+  /**
+   * <p>
+   * Store the given <code>{@link org.quartz.triggers.Trigger}</code>.
+   * </p>
+   *
+   * @param newTrigger The <code>Trigger</code> to be stored.
+   * @param replaceExisting If <code>true</code>, any <code>Trigger</code> existing in the <code>JobStore</code> with the same name & group should be
+   *        over-written.
+   * @throws ObjectAlreadyExistsException if a <code>Trigger</code> with the same name/group already exists, and replaceExisting is set to false.
+   * @see #pauseTriggerGroup(SchedulingContext, String)
+   */
+  @Override
+  public void storeTrigger(OperableTrigger newTrigger, boolean replaceExisting) throws JobPersistenceException {
+
+    TriggerWrapper tw = new TriggerWrapper((OperableTrigger) newTrigger.clone());
+
+    synchronized (lock) {
+      if (wrappedTriggersByKey.get(tw.key) != null) {
+        if (!replaceExisting) {
+          throw new ObjectAlreadyExistsException(newTrigger);
+        }
+
+        removeTrigger(newTrigger.getName());
+      }
+
+      if (retrieveJob(newTrigger.getJobName()) == null) {
+        throw new JobPersistenceException("The job (" + newTrigger.getJobName() + ") referenced by the trigger does not exist.");
+      }
+
+      // add to triggers array
+      wrappedTriggers.add(tw);
+
+      // add to triggers by FQN map
+      wrappedTriggersByKey.put(tw.key, tw);
+
+      if (blockedJobs.contains(tw.jobKey)) {
+        tw.state = TriggerWrapper.STATE_BLOCKED;
+      } else {
+        timeWrappedTriggers.add(tw);
+      }
+    }
+  }
+
+  @Override
+  public boolean removeTrigger(String triggerName) {
+
+    boolean found = false;
+
+    synchronized (lock) {
+      // remove from triggers by FQN map
+      found = (wrappedTriggersByKey.remove(triggerName) == null) ? false : true;
+      if (found) {
+        TriggerWrapper tw = null;
+        // remove from triggers array
+        Iterator<TriggerWrapper> tgs = wrappedTriggers.iterator();
+        while (tgs.hasNext()) {
+          tw = tgs.next();
+          if (triggerName.equals(tw.key)) {
+            tgs.remove();
+            break;
+          }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 278:284, 315:321.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.